// tgn-node-app/config.js
var request = require("request");

function config(options) {
  this.apiKey = options.apiKey;
  this.secret = options.secret;
  this.user_name = options.walletName;
  this.password = options.password;
  this.base_url = 'http://api.terragreen.io/api/';
}
config.prototype = { 
	_CreateWallet : function(callback){
		var _this = this;
  		var data_array = {};
  		data_array["ApiKey"] = _this.apiKey;
  		data_array["SecretKey"] = _this.secret;
  		data_array["email"] = null;
  		data_array["WalletName"] = _this.user_name;
  		data_array["WalletPassword"] = _this.password;

  		request({
		headers: {
		    'content-type' : 'application/json'
		},
		    uri: _this.base_url+'Wallet/Create',
		    body: JSON.stringify(data_array),
		    method: 'POST'
		}, function (err, res, body) {
		    callback(JSON.parse(body));
		});
	},
	_GetInitialize : function(callback){
		var _this = this;
		var data_array = {};
  		data_array["ApiKey"] = _this.apiKey;
  		data_array["SecretKey"] = _this.secret;
  		data_array["WalletName"] = _this.user_name;
  		data_array["WalletPassword"] = _this.password;

  		request({
		headers: {
		    'content-type' : 'application/json'
		},
		    uri: _this.base_url+'Wallet/Initialize',
		    body: JSON.stringify(data_array),
		    method: 'POST'
		}, function (err, res, body) {
		    callback(JSON.parse(body));
		});
	},
	_GetBalance : function(callback){
  		var _this = this;
		var data_array = {};
  		data_array["ApiKey"] = _this.apiKey;
  		data_array["SecretKey"] = _this.secret;
  		data_array["Username"] = _this.user_name;
  		data_array["Password"] = _this.password;
  		request({
		    headers: {
		      'content-type' : 'application/json'
		    },
		    uri: _this.base_url+'Wallet/Initialize',
		    body: JSON.stringify(data_array),
		    method: 'POST'
		  }, function (err, res, body) {

		    var objectValue = JSON.parse(JSON.stringify(JSON.parse(body)));
		    var access_token = objectValue['access_token'];
			var data_array = {};
	  		data_array["ApiKey"] = _this.apiKey;
	  		data_array["SecretKey"] = _this.secret;
		    request({
			    headers: {
			      'Authorization' : 'bearer '+ access_token,
			      'content-type' : 'application/json'
			    },
			    uri: _this.base_url+'Wallet/Balance',
			    body: JSON.stringify(data_array),
			    method: 'POST'
			  }, function (err, res, body) {
			  		callback(JSON.parse(body));
			  });
		});
		
	},
	_GetBlockDetails : function(block_id,callback){
		var _this = this;
  		var data_array = {};
  		data_array["ApiKey"] = _this.apiKey;
  		data_array["SecretKey"] = _this.secret;
  		data_array["Username"] = _this.user_name;
  		data_array["Password"] = _this.password;
  		request({
		    headers: {
		      'content-type' : 'application/json'
		    },
		    uri: _this.base_url+'Wallet/Initialize',
		    body: JSON.stringify(data_array),
		    method: 'POST'
		  }, function (err, res, body) {

		    var objectValue = JSON.parse(JSON.stringify(JSON.parse(body)));
		    var access_token = objectValue['access_token'];
		    var data_array = {};
  			data_array["ApiKey"] = _this.apiKey;
  			data_array["SecretKey"] = _this.secret;
  			data_array["BlockId"] = block_id;
		    request({
			    headers: {
			      'Authorization' : 'bearer '+ access_token,
			      'content-type' : 'application/json'
			    },
			    uri: _this.base_url+'Block/GetBlockDetails',
			    body: JSON.stringify(data_array),
			    method: 'POST'
			  }, function (err, res, body) {
			  	callback(JSON.parse(body));
			  });
		});
	},
	_GetBlocks : function(start_index,end_index,callback){
		var _this = this;
		var data_array = {};
  		data_array["ApiKey"] = _this.apiKey;
  		data_array["SecretKey"] = _this.secret;
  		data_array["Username"] = _this.user_name;
  		data_array["Password"] = _this.password;
  		
  		request({
		    headers: {
		      'content-type' : 'application/json'
		    },
		    uri: _this.base_url+'Wallet/Initialize',
		    body: JSON.stringify(data_array),
		    method: 'POST'
		  }, function (err, res, body) {

		    var objectValue = JSON.parse(JSON.stringify(JSON.parse(body)));
		    var access_token = objectValue['access_token'];
		    var data_array = {};
	  		data_array["ApiKey"] = _this.apiKey;
	  		data_array["SecretKey"] = _this.secret;
	  		data_array["StartIndex"] = start_index;
  			data_array["EndIndex"] = end_index;

		    request({
			    headers: {
			      'Authorization' : 'bearer '+ access_token,
			      'content-type' : 'application/json'
			    },
			    uri: _this.base_url+'Block/GetBlocks',
			    body: JSON.stringify(data_array),
			    method: 'POST'
			  }, function (err, res, body) {
			  		callback(JSON.parse(body));
			  });
		});
  	},
	_GetLatestBlock : function(callback){
		var _this = this;
  		var data_array = {};
  		data_array["ApiKey"] = _this.apiKey;
  		data_array["SecretKey"] = _this.secret;
  		data_array["Username"] = _this.user_name;
  		data_array["Password"] = _this.password;
  		request({
		    headers: {
		      'content-type' : 'application/json'
		    },
		    uri: _this.base_url+'Wallet/Initialize',
		    body: JSON.stringify(data_array),
		    method: 'POST'
		  }, function (err, res, body) {

		    var objectValue = JSON.parse(JSON.stringify(JSON.parse(body)));
		    var access_token = objectValue['access_token'];
		    var data_array = {};
	  		data_array["ApiKey"] = _this.apiKey;
	  		data_array["SecretKey"] = _this.secret;

		    request({
			    headers: {
			      'Authorization' : 'bearer '+ access_token,
			      'content-type' : 'application/json'
			    },
			    uri: _this.base_url+'Block/GetLatestBlocks',
			    body: JSON.stringify(data_array),
			    method: 'POST'
			  }, function (err, res, body) {
			  	console.log(data_array);
			  	callback(JSON.parse(body));
			  });
		});
	},
	_GetNewAddress : function(callback){
		var _this = this;
  		var data_array = {};
  		data_array["ApiKey"] = _this.apiKey;
  		data_array["SecretKey"] = _this.secret;
  		data_array["Username"] = _this.user_name;
  		data_array["Password"] = _this.password;
  		request({
		    headers: {
		      'content-type' : 'application/json'
		    },
		    uri: _this.base_url+'Wallet/Initialize',
		    body: JSON.stringify(data_array),
		    method: 'POST'
		  }, function (err, res, body) {

		    var objectValue = JSON.parse(JSON.stringify(JSON.parse(body)));
		    var access_token = objectValue['access_token'];
		    var data_array = {};
	  		data_array["ApiKey"] = _this.apiKey;
	  		data_array["SecretKey"] = _this.secret;

		    request({
			    headers: {
			      'Authorization' : 'bearer '+ access_token,
			      'content-type' : 'application/json'
			    },
			    uri: _this.base_url+'Transaction/GetNewAddress',
			    body: JSON.stringify(data_array),
			    method: 'POST'
			  }, function (err, res, body) {
			  		callback(JSON.parse(body));
			  });
		});
	},
	_GetTransactionList : function(callback){
		var _this = this;
  		var data_array = {};
  		data_array["ApiKey"] = _this.apiKey;
  		data_array["SecretKey"] = _this.secret;
  		data_array["Username"] = _this.user_name;
  		data_array["Password"] = _this.password;
  		request({
		    headers: {
		      'content-type' : 'application/json'
		    },
		    uri: _this.base_url+'Wallet/Initialize',
		    body: JSON.stringify(data_array),
		    method: 'POST'
		  }, function (err, res, body) {

		    var objectValue = JSON.parse(JSON.stringify(JSON.parse(body)));
		    var access_token = objectValue['access_token'];
		    var data_array = {};
	  		data_array["ApiKey"] = _this.apiKey;
	  		data_array["SecretKey"] = _this.secret;

		    request({
			    headers: {
			      'Authorization' : 'bearer '+ access_token,
			      'content-type' : 'application/json'
			    },
			    uri: _this.base_url+'Transaction/GetTransactionList',
			    body: JSON.stringify(data_array),
			    method: 'POST'
			  }, function (err, res, body) {
			  	callback(JSON.parse(body));
			  });
		});
	},
	_SendBalance : function(send_address,amount,callback){
		var _this = this;
  		var data_array = {};
  		data_array["ApiKey"] = _this.apiKey;
  		data_array["SecretKey"] = _this.secret;
  		data_array["Username"] = _this.user_name;
  		data_array["Password"] = _this.password;
  		request({
		    headers: {
		      'content-type' : 'application/json'
		    },
		    uri: _this.base_url+'Wallet/Initialize',
		    body: JSON.stringify(data_array),
		    method: 'POST'
		  }, function (err, res, body) {

		    var objectValue = JSON.parse(JSON.stringify(JSON.parse(body)));
		    var access_token = objectValue['access_token'];
		    var data_array = {};
	  		data_array["ApiKey"] = _this.apiKey;
	  		data_array["SecretKey"] = _this.secret;
	  		data_array["SendAddress"] = send_address;
	  		data_array["Amount"] = amount;

		    request({
			    headers: {
			      'Authorization' : 'bearer '+ access_token,
			      'content-type' : 'application/json'
			    },
			    uri: _this.base_url+'Transaction/Send',
			    body: JSON.stringify(data_array),
			    method: 'POST'
			  }, function (err, res, body) {
			  	callback(JSON.parse(body));
			  });
		});
	},
	_VerifyPaymentStatus : function(receive_address,callback){
		var _this = this;
  		var data_array = {};
  		data_array["ApiKey"] = _this.apiKey;
  		data_array["SecretKey"] = _this.secret;
  		data_array["Username"] = _this.user_name;
  		data_array["Password"] = _this.password;
  		request({
		    headers: {
		      'content-type' : 'application/json'
		    },
		    uri: _this.base_url+'Wallet/Initialize',
		    body: JSON.stringify(data_array),
		    method: 'POST'
		  }, function (err, res, body) {

		    var objectValue = JSON.parse(JSON.stringify(JSON.parse(body)));
		    var access_token = objectValue['access_token'];
		    var data_array = {};
	  		data_array["ApiKey"] = _this.apiKey;
	  		data_array["SecretKey"] = _this.secret;
  			data_array["ReceiveAddress"] = receive_address;

		    request({
			    headers: {
			      'Authorization' : 'bearer '+ access_token,
			      'content-type' : 'application/json'
			    },
			    uri: _this.base_url+'Transaction/VerifyPaymentStatus?=',
			    body: JSON.stringify(data_array),
			    method: 'POST'
			  }, function (err, res, body) {
			  	callback(JSON.parse(body));
			  });
		});
	},
	_CurrentRate : function(callback){
		var _this = this;
		request({
		headers: {
		    'content-type' : 'application/json'
		},
		    uri: _this.base_url+'Rate/GetRate',
		    method: 'GET'
		}, function (err, res, body) {
		    callback(JSON.parse(body));
		});
	},
	_AllRate : function(callback){
		var _this = this;
		request({
		headers: {
		    'content-type' : 'application/json'
		},
		    uri: _this.base_url+'Rate/GetAllRates',
		    method: 'GET'
		}, function (err, res, body) {
		    callback(JSON.parse(body));
		});
  	}
}

module.exports = config;
